crda/remotematcher.go

```diff
@@ -0,0 +1,386 @@
+//
+// If not passed to NewMatcher, a defaultBatchSize will be used.
+func WithBatchSize(batchSize int) Option {
+	return func(m *Matcher) error {
+		m.batchSize = batchSize
+		return nil
+	}
+}
+
+// Name implements driver.Matcher.
+func (*Matcher) Name() string { return "crda" }
+
+// Maps the crda ecosystem to claircore.Repository.Name.
+func ecosystemToRepositoryName(ecosystem string) string {
+	switch ecosystem {
+	case "maven":
+		return "maven"
+	case "pypi":
+		return "pypi"
+	default:
+		panic(fmt.Sprintf("unknown ecosystem %s", ecosystem))
+	}
+}
+
+// Filter implements driver.Matcher.
+func (m *Matcher) Filter(record *claircore.IndexRecord) bool {
+	if record.Repository == nil {
+		return false
+	}
+	return record.Repository.Name == ecosystemToRepositoryName(m.ecosystem)
+}
+
+// Query implements driver.Matcher.
+func (*Matcher) Query() []driver.MatchConstraint {
+	panic("unreachable")
+}
+
+// Vulnerable implements driver.Matcher.
+func (*Matcher) Vulnerable(ctx context.Context, record *claircore.IndexRecord, vuln *claircore.Vulnerability) (bool, error) {
+	// RemoteMatcher can match Package and Vulnerability.
+	panic("unreachable")
+}
+
+// Config is the configuration accepted by the Matcher.
+//
+// By convention, it's in a map key called "crda".
+type Config struct {
+	URL         string `json:"url" yaml:"url"`
+	Concurrency int    `json:"concurrent_requests" yaml:"concurrent_requests"`
+	Source      string `json:"source" yaml:"source"`
+	Key         string `json:"key" yaml:"key"`
+}
+
+// Configure implements driver.MatcherConfigurable.
+func (m *Matcher) Configure(ctx context.Context, f driver.MatcherConfigUnmarshaler, c *http.Client) error {
+	var cfg Config
+	if err := f(&cfg); err != nil {
+		return err
+	}
+
+	if cfg.Concurrency > 0 {
+		m.requestConcurrency = cfg.Concurrency
+	}
+
+	if cfg.URL != "" {
+		u, err := url.Parse(cfg.URL)
+		if err != nil {
+			return err
+		}
+		m.url = u
+	}
+	if cfg.Source != "" {
+		m.source = cfg.Source
+	}
+	if cfg.Key != "" {
+		m.key = cfg.Key
+	}
+	m.client = c
+
+	return nil
+}
+
+// QueryRemoteMatcher implements driver.RemoteMatcher.
+func (m *Matcher) QueryRemoteMatcher(ctx context.Context, records []*claircore.IndexRecord) (map[string][]*claircore.Vulnerability, error) {
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "crda/Matcher.QueryRemoteMatcher"))
+	zlog.Debug(ctx).
+		Int("records", len(records)).
+		Msg("request")
+
+	// map Packge{name@version} to Packge to associate it with Vulnerability.
+	packageVersionToIndexRecord := make(map[string]*claircore.IndexRecord)
+	key := func(Name, Version string) string {
+		return fmt.Sprintf("%s@%s", Name, Version)
+	}
+	for _, ir := range records {
+		packageVersionToIndexRecord[key(ir.Package.Name, ir.Package.Version)] = ir
+	}
+
+	results := make(map[string][]*claircore.Vulnerability)
+	ctrlC := m.invokeComponentAnalysesInBatch(ctx, records)
+
+	for vrs := range ctrlC {
```

Comment:
This outer for loop can be encapsulated inside the function `m.invokeComponentAnalysesInBatch()`, it means the channel's lifecycle is contained entirely within 1 function.

Reply:
I agree but this was the issue why i chose to not https://github.com/quay/claircore/pull/423#discussion_r680910136